index.js

```diff
@@ -0,0 +1,414 @@
+  // after the transaction succeeded.
+  return paypalPurchase;
+};
+
+/**
+ * Gets the order from PayPal then ensures the client's order
+ * matches the order from PayPal.
+ *
+ * @param {object} options - Options to use.
+ * @param {object} options.order - Client's PayPal order.
+ *
+ * @returns {Promise<object>} Returns the verified PayPal order and purchase.
+ */
+const verifyOrder = async ({order}) => {
+  // An order should never be CREATED
+  // at this stage so let's just cancel it.
+  if(order.status === 'CREATED') {
+    await deleteOrder({order});
+    return {};
+  }
+  if(order.status !== 'COMPLETED') {
+    throw new BedrockError(
+      `Expected PayPal Status COMPLETED got ${order.status}`,
+      Errors.Data);
+  }
+  const {purchase_units: paypalPurchases} = order;
+  const verifiedPurchase = verifyPurchase({paypalPurchases});
+  // Guards against the user re-using a previous order from PayPal.
+  const existingPayments = await paymentService.db.findAll(
+    {query: {service: 'paypal', serviceId: order.id}});
+  // there should be at most 1 existingPayments.
+  if(existingPayments.length > 1) {
+    // This could be problematic as the user whose card was
+    // charged might not get their product.
+    // This does prevent duplicate charges using the same PayPal order id.
+    throw new BedrockError(
+      `More than one Payment found for PayPal order ${order.id}`,
+      Errors.Duplicate
+    );
+  }
+  verifiedPurchase.totalCost = getTotalCost({order});
+  return verifiedPurchase;
+};
+
+/**
+ * Creates a PayPal order server side with no payment
+ * information filled out.
+ *
+ * @param {object} options - Options to use.
+ * @param {Array<object>} options.purchase_units - A purchase_unit.
+ * @param {string} [options.intent = 'CAPTURE'] - The order intent.
+ *
+ * @returns {Promise<object>} The order object.
+ */
+const createGatewayPayment = async ({payment, intent = 'CAPTURE'}) => {
+  const {id, amount: value, currency = 'USD'} = payment;
+  const amount = formatAmount({amount: {currency_code: currency, value}});
+  const purchase_units = [{
+    reference_id: id,
+    amount
+  }];
+  const {brandName, shippingPreference} = config.paypal;
+  const application_context = {
+    brand_name: brandName || 'bedrock-order',
+    shipping_preference: shippingPreference || 'NO_SHIPPING'
+  };
+  logger.debug('ORDER APPLICATION CONTEXT', {application_context});
+  const {headers, api} = await getOptions();
+  const url = `${api}/v2/checkout/orders`;
+  headers['Content-Type'] = 'application/json';
+  const options = {headers};
+  const body = {
+    intent,
+    purchase_units,
+    application_context
+  };
+  const {data} = await axios.post(url, body, options);
+  return data;
+};
+
+const processGatewayPayment = async ({payment}) => {
+  const order = await getOrderFromPayment({payment});
+  const verifiedPurchase = verifyOrder({order});
+  return verifiedPurchase;
+};
+
+module.exports = {
+  type: 'paymentsPayPalPlugin',
```

Comment:
maybe type = 'paymentsPlugin' and id = 'paypal'.  Mainly I think these type of plugins should all have the same type and some other way of differentiating the name.  I don't know what pattern is being used here, but this is what we did elsewhere: https://github.com/digitalbazaar/bedrock-ledger-storage-mongodb/blob/master/lib/index.js#L49

Reply:
Good idea so I did that here: https://github.com/digitalbazaar/bedrock-payment-paypal/pull/2/commits/ded7a74d86b8fe8c87f348c3308380ddbd83b40b